### PR TITLE
Formalizing the JobStore document format as a pydantic model

### PR DESCRIPTION
**Pull request by hrushikesh-s**
## Summary

Include a summary of major changes in bullet points:

- Adding the functionality to parse the JobStore document as a `pydantic model` in place of `dictionary`

## Checklist

Work-in-progress pull requests are encouraged, but please put [WIP] in the pull request
title.

Before a pull request can be merged, the following items must be checked:

- [X] Code is in the [standard Python style](https://www.python.org/dev/peps/pep-0008/).
  The easiest way to handle this is to run the following in the **correct sequence** on
  your local machine. Start with running [black](
  https://black.readthedocs.io/en/stable/index.html) on your new code. This will
  automatically reformat your code to PEP8 conventions and removes most issues. Then run
  [pycodestyle](https://pycodestyle.readthedocs.io/en/latest/), followed by [flake8](
  http://flake8.pycqa.org/en/latest/).
- [X] Docstrings have been added in the[Numpy docstring format](
  https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_numpy.html).
  Run [pydocstyle](http://www.pydocstyle.org/en/2.1.1/index.html) on your code.
- [X] Type annotations are **highly** encouraged. Run [mypy](http://mypy-lang.org/) to
  type check your code.
- [X] Tests have been added for any new functionality or bug fixes.
- [X] All linting and tests pass.

Note that the CI system will run all the above checks. But it will be much more
efficient if you already fix most errors prior to submitting the PR. It is highly
recommended that you use the pre-commit hook provided in the repository. Simply
`cp pre-commit .git/hooks` and a check will be run prior to allowing commits.

### CODE REVIEW

**Comment by hrushikesh-s** [comment]
Hi @utf and @arosen93 ,
Starting a PR to address one of the issues raised in [#374 ]( https://github.com/materialsproject/jobflow/issues/374#issuecomment-1664538475)

**Comment by arosen93** [comment]
Thanks! Linking this to #422 (as a note to future me) in case Pydantic 2 breaks things later.

File: src/jobflow/schemas/job_output_schema.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    hosts: List[str] = Field(
+        None,
+        description="The list of UUIDs of the hosts containing the job.",
+    )
+    name: str = Field(
+        None,
+        description="The name of the job.",
+    )
+
+    @validator("output", pre=True)
```
Comment (utf) [inline src/jobflow/schemas/job_output_schema.py:39]: Hi @hrushikesh-s, please can you update this to Pydantic2 syntax, as discussed here: https://github.com/materialsproject/jobflow/pull/425#discussion_r1332332522

**Comment by hrushikesh-s** [comment]
Hi @utf, the `JobStoreDocument` is now being initialized as a `Pydantic v2` document, and the PR is ready to be merged now!

**Comment by utf** [comment]
Great, thanks for this.

**Comment by mkhorton** [comment]
Thanks for adding the suggested code @hrushikesh-s! Only just saw this was merged, appreciated your efforts on this :)

**Pull request merged**